compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -334,6 +334,8 @@ pub struct InferCtxt<'tcx> {
     pub intercrate: bool,
 
     next_trait_solver: bool,
+
+    pub fulfilled_obligations: Lrc<Lock<Vec<traits::FulfilledObligation<'tcx>>>>,
```

Comment:
Please change this to only be a field of the fulfillment context (and a method on the `Engine` trait to get them. Do not add it to the inference context. That will miss some obligations checked while probing, but given that we also sometimes prove by just using evaluate, this shouldn't change

However, having it on the infcx stores obligations from probes, which can leak inference variables, which is bad:tm: and potentially results in ICE

Reply:
Leaking inference variables is obviously not good. However, there is a motivation for catching obligations from within a probe. Take the following code:

```rust
#![feature(negative_impls)]

use rayon::prelude::*;
use std::sync::MutexGuard;

#[derive(Debug)]
struct NotSend;

impl !Send for NotSend {}
impl !Sync for NotSend {}

fn does_impl<T: IntoParallelIterator>() {}
fn check_impl() { 
    does_impl::<Vec<NotSend>>()
}

fn parallel_do(vs: Vec<NotSend>) {
    // Obligation `Vec<NotSend>: IntoParallelIterator` 
    // happens inside a probe. (At least IIUC)
    vs.into_par_iter();
}

fn main() {}
```

If my understanding of the trait solver is correct, when type-checking the expression `vs.into_par_iter()` the query `Vec<NotSend>: IntoParallelIterator` would happen inside of a probe. For the context of the "trait debugger" we're building, catching this obligation is important so that we can explain the call ambiguity to a user. 

Is there a way to get this obligation without storing the obligations on the `InferCtxt`?